packages/zero-pg/src/schema.ts

```diff
@@ -0,0 +1,82 @@
+import type {ServerSchema} from '../../z2s/src/schema.ts';
+import {formatPg, sql} from '../../z2s/src/sql.ts';
+import type {Schema} from '../../zero-schema/src/builder/schema-builder.ts';
+import type {DBTransaction} from '../../zql/src/mutate/custom.ts';
+
+export type ServerSchemaRow = {
+  schema: string;
+  table: string;
+  column: string;
+  type: string;
+  enum: string;
+};
+
+export async function getServerSchema<S extends Schema>(
+  dbTransaction: DBTransaction<unknown>,
+  schema: S,
+): Promise<ServerSchema> {
+  const serverToClient: Record<string, string> = {};
+  const schemaTablePairs: [string, string][] = [];
+  for (const {name, serverName} of Object.values(schema.tables)) {
+    let schemaTablePair: [string, string] = ['public', serverName ?? name];
+    if (serverName) {
+      const firstPeriod = serverName.indexOf('.');
+      if (firstPeriod > -1) {
+        schemaTablePair = [
+          serverName.substring(0, firstPeriod),
+          serverName.substring(firstPeriod + 1, serverName.length),
+        ];
+      }
+    }
+    schemaTablePairs.push(schemaTablePair);
+    serverToClient[`${schemaTablePair[0]}.${schemaTablePair[1]}`] = name;
+  }
+
+  if (schemaTablePairs.length === 0) {
+    return {}; // No pairs to query for
+  }
+  const inClause = sql.join(
+    schemaTablePairs.map(
+      ([schema, table]) => sql`(${schema}::text, ${table})::text`,
+    ),
+    ',',
+  );
+
+  const query = sql`
+      SELECT
+          c.table_schema::text as schema,
+          c.table_name::text as table,
+          c.column_name::text as column,
+          c.data_type::text AS type,
+          t.typtype = 'e'::text as enum
+      FROM
+          information_schema.columns c
+      JOIN
+          pg_catalog.pg_type t ON c.udt_name = t.typname
+      JOIN
+          pg_catalog.pg_namespace n ON t.typnamespace = n.oid
+      WHERE
+          (c.table_schema, c.table_name) IN (${inClause})
+    `;
+  const {text, values} = formatPg(query);
+  const results: Iterable<ServerSchemaRow> = (await dbTransaction.query(
+    text,
+    values,
+  )) as Iterable<ServerSchemaRow>;
+
+  const serverSchema: ServerSchema = {};
+  for (const row of results) {
+    serverSchema[
+      row.schema === 'public' ? row.table : `${row.schema}.${row.table}`
+    ] = {
+      [row.column]: {
+        type: row.type,
+        isEnum: row.enum.toLowerCase().startsWith('f'),
+      },
+    };
+  }
+
+  // Validate that schema is a valid subset of serverSchema
```

Comment:
nit: // Validate that schema.ts is a valid subset of serverSchema and types are compatible.